tfx/orchestration/experimental/kubernetes/kubernetes_remote_runner.py

```diff
@@ -0,0 +1,190 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Kubernetes TFX runner for out-of-cluster orchestration."""
+
+import absl
+import datetime
+import json
+import time
+from typing import List, Text
+
+from tfx.components.base import base_node
+from tfx.orchestration import pipeline as tfx_pipeline
+from tfx.orchestration.kubeflow import node_wrapper
+from tfx.utils import json_utils, kube_utils
+
+from google.protobuf import json_format
+from kubernetes import client
+
+
+_ORCHESTRATOR_COMMAND = [
+    'python', '-m', 'tfx.orchestration.experimental.kubernetes.orchestrator_container_entrypoint'
+]
+
+def run_as_kubernetes_job(pipeline: tfx_pipeline.Pipeline,
+                          tfx_image: Text) -> None:
+  """Submits and runs a tfx pipeline from outside the cluster.
+
+  Args:
+    pipeline: Logical pipeline containing pipeline args and components.
+  """
+
+  # TODO(ccy): Look for alternative serialization schemes once available.
+  serialized_pipeline = _serialize_pipeline(pipeline)
+  # Extract and pass pipeline graph information which are lost during the
+  # serialization process. The orchestrator container uses downstream_ids
+  # to reconstruct pipeline graph.
+  downstream_ids = json.dumps(_extract_downstream_ids(pipeline.components))
+  arguments = [
+      '--serialized_pipeline',
+      serialized_pipeline,
+      '--downstream_ids',
+      downstream_ids,
+      '--tfx_image',
+      tfx_image,
+  ]
+  batch_api = kube_utils.make_batch_v1_api()
+  job_name = 'Job_' + pipeline.pipeline_info.run_id
+  pod_label = kube_utils.sanitize_pod_name(job_name)
+  container_name = 'pipeline-orchestrator'
+  job = kube_utils.make_job_object(
+      name=job_name,
+      container_image=tfx_image,
+      command=_ORCHESTRATOR_COMMAND + arguments,
+      container_name=container_name,
+      pod_labels={
+          'job-name': pod_label,
+      },
+  )
+  try:
+    batch_api.create_namespaced_job(
+        "default", job, pretty=True)
+  except client.rest.ApiException as e:
+    raise RuntimeError('Failed to submit job! \nReason: %s\nBody: %s' %
+                       (e.reason, e.body))
+
+  # Wait for pod to start.
+  orchestrator_pods = []
+  core_api = kube_utils.make_core_v1_api()
+  start_time = datetime.datetime.utcnow()
+
+  # Wait for the kubernetes job to launch a pod.
+  # This is expected to only take a few seconds.
+  while not orchestrator_pods and (
+      datetime.datetime.utcnow() - start_time).seconds < 300:
```

Comment:
Please extract 300 into a constant `JOB_CREATION_TIMEOUT = 300` at the top of the file, with a comment that it is the timeout for creating a job.

Reply:
Refactored.